include/crocoddyl/core/cost-base.hpp

```diff
@@ -211,37 +241,37 @@ struct CostDataAbstractTpl {
   CostDataAbstractTpl(Model<Scalar>* const model, DataCollectorAbstract* const data)
       : shared(data),
         activation(model->get_activation()->createData()),
+        residual(model->get_residual()->createData(data)),
         cost(Scalar(0.)),
         Lx(model->get_state()->get_ndx()),
         Lu(model->get_nu()),
         Lxx(model->get_state()->get_ndx(), model->get_state()->get_ndx()),
         Lxu(model->get_state()->get_ndx(), model->get_nu()),
-        Luu(model->get_nu(), model->get_nu()),
-        r(model->get_activation()->get_nr()),
-        Rx(model->get_activation()->get_nr(), model->get_state()->get_ndx()),
-        Ru(model->get_activation()->get_nr(), model->get_nu()) {
+        Luu(model->get_nu(), model->get_nu()) {
     Lx.setZero();
     Lu.setZero();
     Lxx.setZero();
     Lxu.setZero();
     Luu.setZero();
-    r.setZero();
-    Rx.setZero();
-    Ru.setZero();
   }
   virtual ~CostDataAbstractTpl() {}
 
+  DEPRECATED("Use residual.r", const VectorXs& get_r() const { return residual->r; };)
```

Comment:
I would copy your comments for more clarity. You have written in https://github.com/loco-3d/crocoddyl/pull/899#discussion_r556545744 :

> Don't remove r, Rx and Ru. Make them references to residual->r, residual->Rx and residual-Ru, and deprecate them. There are users who are using costdata->r values

It is clear to me how to use `Eigen::Ref` for retrieving the values from `residual->XX` (i.e., "make them references of"). What I don't see possible (or reasonable) from your comment is the deprecation of these member objects. Let me explain you this statement. A way of deprecating this object would be
```cpp
DEPRECATED("Use residual.r", Eigen::Ref<VectorXs> r);
```
However, it is not useful as this deprecation code will always print a deprecation message. In short, unfortunately, there is no a reasonable way to deprecate member objects in a class.

On the other hand, I believe, in this case, it is more informative for users to have a compilation error.
Indeed, it should be straightforward for users to realize how the fixing code looks by inspecting the internal code, e.g.,
```cpp
DEPRECATED("Use residual.r", const VectorXs& get_r() const { return residual->r; };)
```

**Closing remarks**:
I believe this is the best approach that we can do. If you have a solution to the deprecation code issue, which I am not aware of, then let me know. Otherwise, I would suggest to mark this conversation as resolved

Reply:
C++ reference would be more efficient than Eigen::Ref here. 
We are already deprecating the constructors of the cost models. Once we finalize on the structure for costmodelresidual, we would add a deprecation to the individual cost models as well.

It is safe to assume that we don't need to deprecate each variable of the costmodel once that happens.
So if we simply replace `VectorXs r` by `VectorXs& r`, I think it should be fine.

Reply:
>  So if we simply replace VectorXs r by VectorXs& r in CostDataAbstract, I think it should be fine.

Yes, using `Eigen::Ref` of this approach is not the issue.

What I mean we could not deprecate this code in c++:
```cpp
VectorXs& r
```
as it will always print a deprecation message, EVEN if the user doesn't include this code.


This is the main issue, we cannot deprecate in c++ attribute in a structure (or public attribute in a class). At least there is no solution that I am aware of, note that I have spent time to find a solution.

Reply:
This is a study limitation of c++ (at least c++98). A workaround might not be trivial, and I am clueless here.